src/Features/Core/Portable/NavigateTo/NavigateToSearcher.cs

```diff
@@ -118,57 +153,65 @@ private async Task SearchCurrentDocumentAsync(CancellationToken cancellationToke
 
             var projectCount = orderedProjects.Sum(g => g.Length);
 
             // We do at least two passes.  One for loaded docs.  One for source generated docs.
-            await _progress.AddItemsAsync(projectCount * 2, cancellationToken).ConfigureAwait(false);
+            await AddProgressItemsAsync(
+                projectCount * ((searchRegularDocuments ? 1 : 0) + (searchGeneratedDocuments ? 1 : 0)),
+                cancellationToken).ConfigureAwait(false);
+
+            if (searchRegularDocuments)
+                await SearchRegularDocumentsAsync(isFullyLoaded, orderedProjects, seenItems, cancellationToken).ConfigureAwait(false);
+
+            if (searchGeneratedDocuments)
+                await SearchGeneratedDocumentsAsync(seenItems, cancellationToken).ConfigureAwait(false);
 
+            // Report a telemetry even to track if we found uncached items after failing to find cached items.
+            // In practice if we see that we are always finding uncached items, then it's likely something
+            // has broken in the caching system since we would expect to normally find values there.  Specifically
+            // we expect: foundFullItems <<< not foundFullItems.
+            if (!isFullyLoaded)
+                Logger.Log(FunctionId.NavigateTo_CacheItemsMiss, KeyValueLogMessage.Create(m => m["FoundFullItems"] = seenItems.Count > 0));
+        }
+
+        private async Task SearchRegularDocumentsAsync(
+            bool isFullyLoaded,
+            ImmutableArray<ImmutableArray<Project>> orderedProjects,
+            HashSet<INavigateToSearchResult> seenItems,
+            CancellationToken cancellationToken)
+        {
             if (!isFullyLoaded)
             {
                 // We need an additional pass to look through cached docs.
-                await _progress.AddItemsAsync(projectCount, cancellationToken).ConfigureAwait(false);
-
+                await AddProgressItemsAsync(orderedProjects.Sum(g => g.Length), cancellationToken).ConfigureAwait(false);
                 await SearchCachedDocumentsAsync(orderedProjects, seenItems, cancellationToken).ConfigureAwait(false);
 
                 // If searching cached data returned any results, then we're done.  We've at least shown some results
                 // to the user.  That will hopefully serve them well enough until the solution fully loads.
                 if (seenItems.Count > 0)
-                {
-                    // Ensure that we actually complete all our workitems so that the progress bar completes.
-                    await _progress.ItemsCompletedAsync(projectCount * 2, cancellationToken).ConfigureAwait(false);
                     return;
-                }
             }
 
             await SearchFullyLoadedProjectsAsync(orderedProjects, seenItems, cancellationToken).ConfigureAwait(false);
-            await SearchGeneratedDocumentsAsync(seenItems, cancellationToken).ConfigureAwait(false);
```

Comment:
since this is streaming, wouldn't we have reported the non source generated results first already?

Reply:
i'm not sure what you mean :)